node.js

```
@@ -0,0 +1,117 @@
+
+/**
+ * This is the entrypoint on node-compatible CommonJS environments.
+ * `asyncLoad` will use `fs.readFile` to load the WASM module.
+ */
+
+const { readFileSync } = require("node:fs");
+const { readFile } = require("node:fs/promises");
+const path = require("node:path");
+const bindings = require("./pkg/matrix_sdk_crypto_wasm_bg.cjs");
+
+const filename = path.join(__dirname, "pkg/matrix_sdk_crypto_wasm_bg.wasm");
+
+// In node environments, we want to automatically load the WASM module
+// synchronously if the consumer did not call `initAsync`. To do so, we install
+// a `Proxy` that will intercept calls to the WASM module.
+bindings.__wbg_set_wasm(
+    new Proxy(
+        {},
+        {
+            get(_target, prop) {
+                const mod = loadModuleSync();
+                return initInstance(mod)[prop];
+            },
+        },
+    ),
+);
+
+/**
+ * Stores a promise which resolves to the WebAssembly module
+ * @type {Promise<WebAssembly.Module> | null}
+ */
+let modPromise = null;
+
+/**
+ * Tracks whether the module has been instantiated or not
+ * @type {boolean}
+ */
+let initialised = false;
+
+/**
+ * Loads the WASM module synchronously
+ *
+ * It will throw if there is an attempt to laod the module asynchronously running
+ *
+ * @returns {WebAssembly.Module}
+ */
+function loadModuleSync() {
+    if (modPromise) throw new Error("The WASM module is being loadded asynchronously but hasn't finished");
+    const bytes = readFileSync(filename);
+    return new WebAssembly.Module(bytes);
+}
+
+/**
+ * Loads the WASM module asynchronously
+ *
+ * @returns {Promise<WebAssembly.Module>}
+ */
+async function loadModule() {
+    const bytes = await readFile(filename);
+    return await WebAssembly.compile(bytes);
+}
+
+/**
+ * Initializes the WASM module and returns the exports from the WASM module.
+ *
+ * @param {WebAssembly.Module} mod
+ * @returns {typeof import("./pkg/matrix_sdk_crypto_wasm_bg.wasm.d")}
+ */
+function initInstance(mod) {
```

```suggestion
function initInstance(mod) {
    if (initialised) {
        // This should be unreachable
        throw new Error("initInstance called twice");
    }
    
```

---

This actually throws if you call `initAsync`, it loads in the background, and then init it synchronously before it finishes; but I think it is fine to throw because the consumer shouldn't do that

---

> This actually throws if you call initAsync, it loads in the background, and then init it synchronously before it finishes;

Will it? In that case, wouldn't `modPromise` be set, so that the synchronous init will fail early with an exception?